cmd/up.go

```diff
@@ -819,3 +840,222 @@ func setupDotfiles(
 
 	return nil
 }
+
+// setupGPGAgent will forward a local gpg-agent into the remote container
+// this works by
+//
+// - stopping remote gpg-agent and removing the sockets
+// - exporting local public keys and owner trust
+// - importing those into the container
+// - ensuring the gpg-agent is stopped in the container
+// - starting a reverse-tunnel of the local unix socket to remote
+// - ensuring paths and permissions are correctly set in the remote
+//
+// this procedure uses some shell commands, in order to batch commands together
+// and speed up the process. this is ok as remotes will always be linux workspaces
+func setupGPGAgent(
+	client client2.BaseWorkspaceClient,
+	devPodConfig *config.Config,
+	log log.Logger,
+) error {
+	execPath, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	writer := log.Writer(logrus.InfoLevel, false)
+
+	sshCmdArgs := []string{
+		"ssh",
+		"--agent-forwarding=true",
+		"--start-services=false",
+		"--context",
+		client.Context(),
+		client.Workspace(),
+		"--log-output=raw",
+	}
+
+	if log.GetLevel() == logrus.DebugLevel {
+		sshCmdArgs = append(sshCmdArgs, "--debug")
+	}
+
+	log.Debugf("Initializing gpg-agent forwarding")
+	// Check if the agent is running in the workspace already.
+	//
+	// this cose is inspired by https://github.com/coder/coder/blob/main/cli/ssh.go
+	killRemoteAgent := append(sshCmdArgs, "--command")
+	killRemoteAgent = append(killRemoteAgent, `sh -c '
+set -e
+set -x
+agent_socket=$(gpgconf --list-dir agent-socket)
+echo $agent_socket
+if [ -S $agent_socket ]; then
+  echo agent socket exists, attempting to kill it >&2
+  gpgconf --kill gpg-agent
+  rm -f $agent_socket
+  sleep 1
+fi
+
+test ! -S $agent_socket
+'`)
```

Comment:
Instead of having this bash script, can we create a devpod command for this, e.g. `devpod agent workspace start-gpg-agent`

Reply:
Done